### ISSUE TITLE
RangeError: Offset is outside the bounds of the DataView

### ISSUE THREAD

**Issue by jsuvanto** [xeokit/xeokit-sdk#184]
**Summary:**
I get the error `RangeError: Offset is outside the bounds of the DataView` when running xeokit-sdk on my React app.

**Expected behaviour:**
No errors, display Schependomlaan model in browser window

**Actual behaviour:**
I get the following messages in browser console:
```
[xeokit viewer 1]: Installed plugin: BIMServerModels 
[default] Login successful                         
Connecting websocket
[default] Getting model data...

// At this point the canvas is rendered as expected, with a loading spinner visible.
// After a few seconds the log continues with the following errors:

[xeokit plugin BIMServerModels]: data does not start with BGS ()
[xeokit plugin BIMServerModels]: Unimplemented protocol version
[xeokit plugin BIMServerModels]: data does not start with BGS ( /* approximately 173 kB of data */)
Uncaught RangeError: Offset is outside the bounds of the DataView
    at DataView.getInt8 (<anonymous>)
    at DataInputStreamReader.readByte (DataInputStreamReader.js:65)
    at processMessage (BIMServerPerformanceGeometryLoader.js:32)
    at BIMServerPerformanceGeometryLoader.process (BIMServerPerformanceGeometryLoader.js:54)
    at Scene.<anonymous> (BIMServerLoaderPlugin.js:367)
    at Scene.fire (Component.js:321)
    at fireTickEvents (core.js:251)
    at frame (core.js:215)
```
The messages "data does not start with BGS" and "Unimplemented protocol version" refer to `index.js:1375`.

The final error message refers to `xeokit-sdk/src/plugins/BIMServerLoaderPlugin/lib/DataInputStreamReader.js:65`:
```javascript
  64 | readByte() {
> 65 |   var value = this.dataView.getInt8(this.pos);
  66 | ^ this.pos += 1;
  67 |   return value;
  68 | }
```

**Steps to reproduce:**
1. Install BIMserver 1.5.172 on local machine.
2. Start server (run `bimserverjar-1.5.172.jar`)
3. Go to http://localhost:8082/ (default port for BIMserver)
4. Install the Schependomlaan model as instructed by BIMserver
5. Create a new React app
6. Use provided code in a React app (see below)
7. Start React app (run `yarn start`)
8. Browse to http://localhost:3000/ (default port for React app)

**Provided code:**

BimServerViewer.js:
```javascript
import React from 'react';
import { Viewer } from "xeokit-sdk/src/viewer/Viewer";
import { BIMServerLoaderPlugin } from 'xeokit-sdk/src/plugins/BIMServerLoaderPlugin/BIMServerLoaderPlugin';
import BimServerClient from 'xeokit-sdk/src/plugins/BIMServerLoaderPlugin/BIMServerClient/bimserverclient';

export default class BimServerViewer extends React.Component {
  

    componentDidMount() {      

        const { address, username, password, poid, canvasId } = this.props;

        const viewer = new Viewer({
            canvasId: canvasId
        })
        
        const bimServerClient = new BimServerClient(address);

        const loader = new BIMServerLoaderPlugin(viewer, {
            bimServerClient: bimServerClient
        });

        bimServerClient.init(() => {
            bimServerClient.login(username, password, () => {
                
                bimServerClient.call("ServiceInterface", "getProjectByPoid", {
                    poid: poid
                }, (project) => {

                    const model = loader.load({
                        id: "myModel",
                        poid: poid,
                        roid: project.lastRevisionId,
                        schema: project.schema,
                        scale: [0.001, 0.001, 0.001],
                        rotation: [-90, 0, 0],
                        edges: true
                    });

                    model.on("loaded", () => {
                        viewer.cameraFlight.jumpTo(model);
                    });

                    model.on("error", (errMsg) => {
                        console.error("Loading failed: " + errMsg);
                    });

                });
            }, (error) => {
                    console.error("Error logging to BIMserver: " + error.message)
            });
        });
    }

    render() {
        const {
            canvasId,
            width,
            height
        } = this.props;

        return(
            <canvas
                id={canvasId}
                width={width}
                height={height} />
        )
    }
}
```

App.js (extract):
```javascript
import React from 'react';
import BimServerViewer from "./BimServerViewer";

function App() {
    return (
        <BimServerViewer 
            address="http://localhost:8082/"
            username="redacted"
            password="redacted"
            poid="131073" // this is the poid of the Schependomlaan project
            canvasId="myCanvas" />
    )
}

export default App;
```

**Comment by xeolabs** [comment]
Yeah this is expected - currently BIMServerLoaderPlugin only supports BIMServer v1.5.120.

It's on the near-term xeokit roadmap to support 1.5.172, but for now I recommend you use v1.5.120, with IFC schema ifc2x3tc1.

**Comment by xeolabs** [comment]
Thanks for the perfect issue ticket BTW.

**Comment by jsuvanto** [comment]
I thought it would be an incompatibility issue with the version of BIMserver. The error message "Unimplemented protocol version" was slightly misleading, though: I didn't expect the protocol to be different. A more precise error message would be much appreciated.

I will try this with BIMserver 1.5.120 next week and let you know if it worked.

**Comment by jsuvanto** [comment]
Is there any news regarding support for newer versions of bimserver?

**Comment by mantinha** [comment]
I had a similar problem in another situation using BIMserver 1.5.180 and an ifc file generated in Freecad in the ifc2x3tc1 schema

The code used was the same as the tutorial, I just changed the access link, login and password.
https://github.com/xeokit/xeokit-sdk/wiki/Viewing-Models-from-BIMServer

![image](https://user-images.githubusercontent.com/25441376/71895972-3de81580-3131-11ea-9c81-a769cfbc2307.png)
---------------------------------------------------------------------------------------------------------
![image](https://user-images.githubusercontent.com/25441376/71896683-39246100-3133-11ea-82ce-6b27ced71299.png)

**Comment by xeolabs** [comment]
BIMServerLoaderPlugin currently only works with BIMserver 1.5.120 - see documentation: https://github.com/xeokit/xeokit-sdk/wiki/Viewing-Models-from-BIMServer 

Unfortunately the BIMServer data binary format is undocumented, and the BIMServer code itself is difficult to comprehend, hence the difficulty so far in upgrading to the latest version. 

A free xeokit [Single Website Commercial License](https://xeokit.github.io/xeokit-licensing/) to anyone contributing an upgrade to the latest BIMServer binary format!

**Comment by nissimpradhan** [comment]
@xeolabs any pointers on where you found the info for the binary data format for BIMServer 1.5.120? i want to try to do this, but i'll need help form you.

**Comment by xeolabs** [comment]
@nissimpradhan best approach would be to reverse-engineer the latest version of BIMSurfer, the WebGL2-based viewer that works with BIMServer: https://github.com/opensourceBIM/BIMsurfer

One thing - the new BIMSurfer is using spatial queries to load on-demand the portions of the model as they become visible within the 3D view volume. I believe that there is some sort of mode where BIMSurfer can load the whole model, regardless of what portion is currently visible. For xeokit, we need to load the whole model immediately, so we wouldn't be using that load-portion-on-demand-as-visible mechanism.

I'm thinking the process would be something like:

1. get boundary of whole model, then
2. load objects that are within that boundary. 

It's possible that the BIMServer binary format hasn't even changed much, if you're lucky. Perhaps start by testing the current xeokit BIMServerLoaderPlugin against the latest BIMServer, then take it from there. 
 
Good luck!

**Comment by mantinha** [comment]
Is there any xeokit configuration where I can change webgl 1 to webgl2?

**Comment by nissimpradhan** [comment]
@xeolabs Hi! It's been a while... I have some good news! And you better get that commercial license to me :D
I only had time recently to look into this issue. 

I compared a working version of [BIMVie.ws](https://github.com/opensourceBIM/bimvie.ws) with [first example](https://github.com/xeokit/xeokit-sdk/wiki/Viewing-Models-from-BIMServer) ... more specifically 

- BIMServerPerformanceGeometryLoader.js in Xeokit with
- GeometryLoader.js in BIMVie.ws

After investigating quite deep into how a model is loaded, i found out that ( in the section where binary data is read) the newer version has added binary Long after the first binary Long bit .. and this new Long data denotes a message Type. So because a new bits of data were added to the binary... Xeokit couldn't parse the data properly.

Screenshot: process function in BIMServerPerformanceGeometryLoader.js in Xeokit
![image](https://user-images.githubusercontent.com/10406305/75789536-23659d80-5d6a-11ea-9a74-9c797fb1ac61.png)

Screenshot: process function in GeometryLoader.js in BIMVie.ws
![image](https://user-images.githubusercontent.com/10406305/75789640-4b550100-5d6a-11ea-87f3-b880e2c09582.png)

The blue line is missing in Xeokit

After adding the code to read a binary Long and then adding Protocol 20 ( which seems to be the version that BIMServer 1.5.180 is using) to the list of supported protocols in the if statement, The model seems to render correctly. I will make a pull request request now.
